server/controllers/thread.controller.js

```diff
@@ -22,4 +23,67 @@ function get(req, res, next) {
         .catch(next);
 }
 
-export default { get };
+function list(req, res, next) {
+    const queryObject = {
+        raw: true,
+        where: {},
+    };
+
+    if (req.query.limit) {
+        queryObject.limit = req.query.limit;
+    }
+
+    if (req.query.offset) {
+        queryObject.limit = req.query.limit;
+    }
+
+    const senders = [Sequelize.fn('ARRAY_AGG', Sequelize.fn('DISTINCT', Sequelize.col('from'))), 'senders'];
+    const originalMessageId = 'originalMessageId';
+    const mostRecent = [Sequelize.fn('MAX', Sequelize.col('createdAt')), 'mostRecent'];
+    const count = [Sequelize.fn('COUNT', Sequelize.col('id')), 'count'];
+    const archived = [Sequelize.fn('bool_or', Sequelize.col('isArchived')), 'archived'];
+    const unread = [Sequelize.fn('bool_and', (Sequelize.literal('CASE WHEN "readAt" IS NULL THEN true ELSE false END'))), 'unread'];
+    queryObject.group = 'originalMessageId';
+    queryObject.attributes = [senders, originalMessageId, mostRecent, count, archived, unread];
+
+    Message.scope({ method: ['findAllForUser', req.user] })
+        .findAll(queryObject)
+        .then((threadsResponse) => {
+            Message.scope({ method: ['findAllForUser', req.user] })
+            .findAll({ raw: true })
+            .then((allMessages) => {
+                const threads = threadsResponse;
+                let i;
+                let j;
+                let maxReadDate;
+                let minUnreadCreatedDate;
+                let minSubjectDate;
+                for (i = 0; i < threads.length; i++) { // eslint-disable-line no-plusplus
+                    threads[i].count = parseInt(threads[i].count, 10);
+                    for (j = 0; j < allMessages.length; j++) { // eslint-disable-line no-plusplus
+                        if (threads[i].unread) {
```

Comment:
All of these nested if/else clauses should be refactored out into functions. When reading the main method I should be able to logically understand what is going on by reading a sequence of method names.

For example, `minUnreadCreatedDate > allMessages[j].createdA` can be placed into a method with a name that explains what this check is for. Then place that into another method. If there are two `if`s in a branch you should start breaking out methods.